### ISSUE TITLE
Customizing AutoFixure using FromSeed Causes Exception

### ISSUE THREAD

**Issue by NArnott** [AutoFixture/AutoFixture#467]
Given two classes:

    class Foo
    {
        ...
    }

    class Bar
    {
        public Foo FooBar { get; set; }
    }

I have set up the following test:

    void Test()
    {
        var fixture = new Fixture();

        fixture.Customize<Foo>(x => x.FromSeed(TestFooFactory));

        var fooWithoutSeed = fixture.Create<Foo>();
        var fooWithSeed = fixture.Create<Foo>(new Foo());

        var bar = fixture.Create<Bar>(); //error occurs here
    }
        
    Foo TestFooFactory(Foo seed)
    {
        //do something with seed...

        return new Foo();
    }

I can create `Foo` objects directly with and without seed values without any problem.  But once I try to create a `Bar` object that has a `Foo` property, I get an `ObjectCreationException`:

> The decorated ISpecimenBuilder could not create a specimen based on the request: Foo. This can happen if the request represents an interface or abstract class; if this is the case, register an ISpecimenBuilder that can create specimens based on the request. If this happens in a strongly typed Build<T> expression, try supplying a factory using one of the IFactoryComposer<T> methods.

I'd expect `TestFooFactory` to get passed a `null` seed value during the creation of `Bar`, just as when I created `Foo` without a seed value.

In my real-world scenario, I want to customize how AutoFixture would use seeded values for certain objects when I pass seeded values in, but I still want AutoFixture to default to normal behavior if no seed is provided.

**Comment by ploeh** [comment]
Crossposted on Stack Overflow: http://stackoverflow.com/q/33635042/126014

**Comment by ecampidoglio** [comment]
### tl;dr;

The problem is that the factory function specified through `IFactoryComposer<T>.FromSeed` becomes exclusive for requests of type `T` but is only allowed to handle _seeded requests_, which isn't the case when populating properties and fields.

### Background

When customizing a type with the `IFactoryComposer<T>.FromSeed` method, AutoFixture inserts a specimen builder called [`SeededFactory<T>`](https://github.com/AutoFixture/AutoFixture/blob/master/Src/AutoFixture/Kernel/SeededFactory.cs) into the pipeline. The responsibility of `SeededFactory<T>` is to look for `SeededRequest` objects whose requested type matches `T`. If it finds one, it relays the requests to the custom factory function specified through `IFactoryComposer<T>.FromSeed`. All other requests are handled by returning a `NoSpecimen`.

When AutoFixture populates the members of a class (such as properties and fields) it _unwraps_ a `SeededRequest` into a request for the specific type `T`. Since those requests aren't handled by the `SeededFactory<T>` it returns a `NoSpecimen`, which is ultimately converted into a `ObjectCreationException`.

### Possible Solution

I found that modifying the `SeededFactory<T>` class to [relay even _non-seeded requests_](https://github.com/AutoFixture/AutoFixture/blob/master/Src/AutoFixture/Kernel/SeededFactory.cs#L47-L51) to the custom factory function by passing the default value or `T` solves the problem:

```diff
 var seededRequest = request as SeededRequest;

 if (seededRequest == null)
 {
-    return new NoSpecimen(request);
+    return this.create(default(T));
 }
```

Since this is a change in behavior, [these tests](https://github.com/AutoFixture/AutoFixture/blob/master/Src/AutoFixtureUnitTest/Kernel/SeededFactoryTest.cs#L44-L74) would no longer be valid and should be replaced by new ones. I'm not sure whether this is the _right_ answer, but at least it's a possibility. What do you guys think?

### Compatibility

Regardless of which solution we choose, it's going to be a _breaking change_ so it should be done in the `v4` branch.

**Comment by ploeh** [comment]
I tried to reduce the reported problem to this:

```C#
[Fact]
public void FromSeedRepro()
{
    var fixture = new Fixture();
    fixture.Customize<Version>(x => x.FromSeed(s => s));
    var v1 = fixture.Create<Version>();
    var v2 = fixture.Create<Version>(new Version(2, 0));
    var ph = fixture.Create<PropertyHolder<Version>>();
}
```

The exception thrown looks like it's exactly the same exception as the one thrown by the code in http://stackoverflow.com/q/33619619/126014, for which I've recently submitted #470.

I wonder if #470 would also solve this issue...

**Comment by ecampidoglio** [comment]
I run [that test](https://github.com/AutoFixture/AutoFixture/issues/467#issuecomment-156227595) separately in #469 and #470. It fails in both of them but with different exceptions.

In #470 it fails with an `ObjectCreationException` while in #469 I got an `InvalidOperationException`:

> System.InvalidOperationException : The specimen returned by the decorated ISpecimenBuilder is not compatible with System.Version.

The reason why it fails in #469 is that the factory function `s => s` returns `null` when AutoFixture creates a value for the property `PropertyHolder<Version>.Property`. In #470 the factory function is never invoked.

However, the repro for this particular issue is slightly different:

```csharp
[Fact]
public void CustomizeFromSeedWillPopulatePropertyOfSameType()
{
    // Fixture setup
    var fixture = new Fixture();
    var seed = new Version();
    // Exercise system
    fixture.Customize<Version>(c => c.FromSeed(s => seed));
    // Verify outcome
    Assert.Equal(seed, fixture.Create<PropertyHolder<Version>>().Property);
}
```

In this case the factory function `s => new Version()` always returns a fixed seed value.

When I run it in #470 it fails with the same `ObjectCreationException` because the factory function is never invoked. In #469 it passes.

Now, this is where things get interesting. If I merge both #469 and #470 then _both tests pass_.

I added them in 378991f for reference.
